File: algorithms/130.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
private:
    vector<int> dx = {1, -1, 0, 0}, dy = {0, 0, 1, -1};
    void dfs(vector< vector<char> > &board, int x, int y)
    {
        board[x][y] = '!';
        for ( int i = 0; i < 4; ++i )
            if ( x+dx[i] > 0 && x+dx[i] < (int)board.size()-1 &&
                 y+dy[i] > 0 && y+dy[i] < (int)board[0].size()-1 &&
                 board[x+dx[i]][y+dy[i]] == 'O' )
                dfs(board, x+dx[i], y+dy[i]);
        return;
    }
public:
    void solve(vector< vector<char> >& board)
    {
        if ( board.size() <= 2 || board[0].size() <= 2 )
            return;
        int m = board.size(), n = board[0].size();
        for ( int i = 0; i < m; ++i )
        {
            if ( board[i][0] == 'O' )
                dfs(board, i, 0);
            if ( board[i][n-1] == 'O' )
                dfs(board, i, n-1);
        }
        for ( int i = 0; i < n; ++i )
        {
            if ( board[0][i] == 'O' )
                dfs(board, 0, i);
            if ( board[m-1][i] == 'O' )
                dfs(board, m-1, i);
        }
        for ( int i = 0; i < m; ++i )
            for ( int j = 0; j < n; ++j )
                if ( board[i][j] == '!' )
                    board[i][j] = 'O';
                else
                    board[i][j] = 'X';
        return;
    }
};

int main()
{
    return 0;
}

File: algorithms/python3/130.py
class Solution:
    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        if len(board) == 0:
            return
        m, n = len(board), len(board[0])
        visited = [[False] * n for i in range(m)]
        f = lambda x, y: board[x][y] == 'O' and not visited[x][y]
        for i in range(m):
            if f(i, 0):
                self.helper(i, 0, board, visited)
            if f(i, n-1):
                self.helper(i, n-1, board, visited)
        for i in range(n):
            if f(0, i):
                self.helper(0, i, board, visited)
            if f(m-1, i):
                self.helper(m-1, i, board, visited)
        for i in range(m):
            for j in range(n):
                if f(i, j):
                    board[i][j] = 'X'

    def helper(self, i, j, board, visited):
        m, n = len(board), len(board[0])
        visited[i][j] = True
        f = lambda x, y: board[x][y] == 'O' and not visited[x][y]
        if i-1 >= 0 and f(i-1, j):
            self.helper(i-1, j, board, visited)
        if i+1 < m and f(i+1, j):
            self.helper(i+1, j, board, visited)
        if j-1 >= 0 and f(i, j-1):
            self.helper(i, j-1, board, visited)
        if j+1 < n and f(i, j+1):
            self.helper(i, j+1, board, visited)
